Use compact output for all types when showing arrays

nalimilan opened the pull request:
Any custom type can now declare whether it supports compact output,
in which case array printing will include type information only once
using a prefix.

Also fix a reference to showlimited() -> showcompact().

Cf. https://github.com/JuliaLang/julia/pull/15928#discussion_r60117681

JeffBezanson commented:
It's not clear to me that this is needed. Could we just print all array elements in compact format? If a type doesn't support a special compact format, it can simply ignore the flag in its `show` code and you automatically get non-compact output in that case.

nalimilan commented:
@JeffBezanson I wanted to do that too at first, but then I changed my mind. The problem is that if a type does not provide a compact representation (one without type information), the prefix providing the type will be redundant, taking even more space than the non-compact output. That said, we could mandate that types implement a compact representation if they want a correct display. That would certainly make things simpler.

I also don't understand what the `isa(e,DataType) && e === e.name.primary` is here for. It would be good for somebody else to check it.

JeffBezanson commented:
I can live with that. I think it's better to have slightly ugly output in some cases than to make people jump through hoops like this. Part of the issue is that the `limit_output` export already bothers me. There are way too many exports and things to define around printing, and we have to rein in the complexity somehow.

The `DataType` test is used to print `Array` as `Array` rather than `Array{T,N}`.

nalimilan commented:
I'm fine with always using the compact representation and see how it goes.

JeffBezanson commented:
Ok. I think something we really need is to keep working on the plan described starting here: https://github.com/JuliaLang/julia/issues/14052#issuecomment-168123944
We need to add an enum for output verbosity.

nalimilan commented:
I've updated the PR to remove the condition altogether. This is consistent with changes planned in #14052 and indeed makes the system a bit simpler.

I think this makes sense in general, but this raises questions for some rare cases, in particular complex objects. For example, a pathological case unlikely to be used much in practice:
```julia
julia> showcompact([IOBuffer(),IOBuffer()])
Base.AbstractIOBuffer{Array{UInt8,1}}[IOBuffer(data=UInt8[...], readable=true, writable=true, seekable=true, append=false, size=0, maxsize=Inf, ptr=1, mark=-1),IOBuffer(data=UInt8[...], readable=true, writable=true, seekable=true, append=false, size=0, maxsize=Inf, ptr=1, mark=-1)]
```

Though this could even be considered as an improvement, since this kind of array was already quite hard to read without its type prefix. With the prefix, at least you immediately know what the array contains.

nalimilan committed:
Use compact output for all types when showing arrays

This gets rid of the hard-coded list of types for which the compact
representation is used: all types should implement a compact output,
stripping all type information redundant with the array prefix.
We still keep a list of types for which no prefix is needed since
these are created by default for a given syntax.

Also improve docs and fix a reference to showlimited() -> showcompact().

nalimilan commented:
A test I needed to change to fix a failure illustrates better the problem:
```julia
immutable Foo2517; end
@test repr(Array(Foo2517,1)) == "Foo2517[Foo2517()]"
```

Without this PR, it was `[Foo2517()]`. We could recommend implementing `show` to return `()` when `limited=true`, in which case it would give `Foo2517[()]`.

For more interesting types with e.g. two integer fields, this would give ``Bar[(3, 5), (6, 7)]`` instead of currently ``[Bar(3, 5), Bar(6, 7)]``. Makes sense to me, but do we all agree on this?

JeffBezanson commented:
I think this is fine. `Foo[()]` doesn't really work since it implies that the empty tuple can be converted to a `Foo`, which is not necessarily true.

JeffBezanson commented:
Oh, also, now that `[a, b]` doesn't concatenate, I think we can always skip the type prefix if every element is of the same type.

nalimilan commented:
> Oh, also, now that [a, b] doesn't concatenate, I think we can always skip the type prefix if every element is of the same type.

No, the type prefix is very useful to save space for types which support a compact output skipping the type. For example, an array of `Nullable` can be written `Nullable{Int}[1, 2, 3]` instead of `[Nullable(1), Nullable(2), Nullable(3)]`.

JeffBezanson commented:
How do you know which types can do that?

nalimilan commented:
That's the whole issue. My original PR allowed types to provide a `limit_output` method to indicate this. OTC, the new simpler PR always prints the type prefix, which is simpler (and give more systematic print results) but will in some cases lead to printing a redundant prefix. I'm fine with either solution.

JeffBezanson commented:
Rebased. Should be good to merge now.

nalimilan commented:
AppVeyor failure on 64-bit seems unrelated.
